src/net.cpp

```diff
@@ -1741,6 +1753,79 @@ int CConnman::GetExtraOutboundCount()
     return std::max(nOutbound - nMaxOutbound, 0);
 }
 
+void CConnman::AssignNewDandelionDestination(CNode* from)
+{
+    unsigned min_use{std::numeric_limits<unsigned>::max()};
+    int max_select = 0;
+    for (DandelionPeer& peer : m_nodes_dandelion) {
+        // Skip peers that disconnected and have been replaced
+        if (!peer.first) continue;
+
+        // But consider at most MAX_OUTBOUND_DANDELION peers
+        ++max_select;
+        if (max_select > MAX_OUTBOUND_DANDELION) break;
+
+        // Src and dst must be different
+        if (from->GetId() == peer.first->GetId()) continue;
+
+        // This is the best candidate so far
+        if (min_use > peer.second) {
+            min_use = peer.second;
+            from->m_dandelion_destination = &peer;
+            break; // They are sorted by use count, so the first one is one of the best ones
+        }
+    }
+    if (from->m_dandelion_destination) {
+        // Increase use count
+        ++(from->m_dandelion_destination->second);
+    }
+}
+
+void CConnman::ThreadShuffleDandelionDestinations()
+{
+    while (!interruptNet) {
+        auto time_micros = GetTimeMicros();
+        time_micros = PoissonNextSend(time_micros, INTERVAL_DANDELION) - time_micros;
+
+        LogPrint(BCLog::DANDELION, "Dandelion: Shuffle Dandelion Destinations.\n");
+        {
+            LOCK(cs_dandelion_peers);
+
+            std::vector<CNode*> all_nodes;
+            ForEachNode([&](CNode* node) { all_nodes.push_back(node); });
+
+            // Collect all potential dandelion nodes
+            m_nodes_dandelion.clear();
+            for (CNode* node : all_nodes) {
+                // Ignore inbound nodes
+                if (node->fInbound) continue;
```

Comment:
Why ignore inbound nodes? Does this mean non-listening dandelion nodes send only their own transactions and that the peer who receives such will know with undeniable certainty that the non-listening node initiated the transaction?

Reply:
Inbound nodes are much more likely to be snooping peers.

Reply:
While that is true, aren't there solutions that do not make dandelion break firewalled nodes' privacy? For example, send only to outgoing peers if you originated the transaction, but send to incoming as well for forwarded txes? Perhaps weigh the outgoing peers more heavily than incoming peers when selecting a random peer to forward to (either by assigning a weight to individual peers based on incoming/outgoing status, or by first randomly selecting the outgoing or incoming pool with a fixed weighting before selecting a peer from whichever pool).